### PR TITLE
Fix custom float dtype f-string formatting truncating exponent notation

### LINKED ISSUE #341: Bug: `jnp.bfloat16` f-string formatting truncates exponent
#### Description:
When converting a floating-point number (e.g., `1e-6`) to `jnp.bfloat16` and then printing it using an f-string with a format specifier (e.g., `:.8`), the resulting scientific notation output is truncated, specifically missing the exponent part.

#### Steps to Reproduce:
Run
   ```python
   import jax.numpy as jnp
   ja = jnp.array(1e-6, dtype=jnp.bfloat16)
   print(f'{ja:.8}')
   ```

#### Observed Behavior:
The output truncates the exponent part of the scientific notation. For example, the observed output is:
```
9.98378e
```

#### Expected Behavior:
The output should display the full scientific notation, including the exponent, even if there is some precision loss due to the `bfloat16` conversion. For example, an expected output (after precision loss) would be:
```
9.983780e-7
```

### System info (python version, jaxlib version, accelerator, etc.)

*   JAX version:  0.7.2
*   Python version: 3.12

### PR DESCRIPTION
**Pull request by xhmelon**
This PR fixes #341.

Formatting custom float scalars (e.g. `bfloat16`) would fallback to `np.generic.__format__` and be truncated by string length.

```bash
python -c "from ml_dtypes import bfloat16; print(f'{bfloat16(1e-6):.8}')"
```
> `9.98378e`  — exponent "-07" truncated

This PR adds a `__format__` method to all custom float types that:
1. Rounds the value to 6 significant digits, which matches the default precision of `std::ostringstream` from `PyCustomFloat_Str`/`Repr` methods.
2. Delegates to `float.__format__` for the actual formatting.

After the fix:
```bash
python -c "from ml_dtypes import bfloat16; print(f'{bfloat16(1e-6):.8}')"
```
> `9.98378e-07`

```bash
python -c "from ml_dtypes import bfloat16; print(f'{bfloat16(1e-6):.8e}')"
```
> `9.98378000e-07` — padding digits beyond precision with zeros

### CODE REVIEW

**Comment by xhmelon** [comment]
Hi @jakevdp, would you please activate CI workflows and review this PR?

File: ml_dtypes/_src/custom_float.h [hunk 1]
```diff
@@ -353,6 +353,38 @@ Py_hash_t PyCustomFloat_Hash(PyObject* self) {
...
+  PyObject* format_spec = nullptr;
+  if (!PyArg_ParseTuple(args, "U", &format_spec)) {
+    return nullptr;
+  }
+
+  T x = reinterpret_cast<PyCustomFloat<T>*>(self)->value;
+  // Round to 6 significant digits to match PyCustomFloat_Str/Repr, which
+  // use std::ostringstream with its default precision of 6. This avoids
+  // exposing false precision from the float64 expansion.
+  char buf[14];  // max %.6g output: "-9.98378e+38" + '\0' = 14
```
Comment (hawkinsp) [inline ml_dtypes/_src/custom_float.h:367]: I'm finding this part strange. For any floating point type, numpy appears to do the following: https://github.com/numpy/numpy/blob/f105cf2d7c20c9829b431c2ae0cdb1f07efaccf2/numpy/_core/src/multiarray/scalartypes.c.src#L633

i.e., convert the type to a python float using nb_format and then call PyObject_Format on the result.

Is there a reason for us to deviate here?

**Review by hawkinsp — CHANGES_REQUESTED** [review]
Thanks the the PR!

File: ml_dtypes/_src/custom_float.h (hunk 1 above)
Comment (hawkinsp) [reply ml_dtypes/_src/custom_float.h:367]: What do you think?

**Comment by hawkinsp** [comment]
Closing in favor of https://github.com/jax-ml/ml_dtypes/pull/387, which fixes this as I suggest above and for all the types we support not just floats.

**Comment by hawkinsp** [comment]
Thanks for the PR!
